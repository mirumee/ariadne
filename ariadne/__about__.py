__version__ = "0.0.1.dev0"  # This is overwritten by Hatch in CI/CD, don't change it.
