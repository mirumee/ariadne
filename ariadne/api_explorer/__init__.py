from .api_explorer import APIExplorer, APIExplorerHttp405
from .graphiql import APIExplorerGraphiQL
from .playground import APIExplorerPlayground

__all__ = [
    "APIExplorer",
    "APIExplorerGraphiQL",
    "APIExplorerHttp405",
    "APIExplorerPlayground",
]
