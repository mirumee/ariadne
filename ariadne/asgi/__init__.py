from ..types import (
    Extensions,
    MiddlewareList,
    Middlewares,
    OnComplete,
    OnConnect,
    OnDisconnect,
    OnOperation,
    Operation,
    WebSocketConnectionError,
)
from .graphql import GraphQL

__all__ = [
    "Extensions",
    "GraphQL",
    "MiddlewareList",
    "Middlewares",
    "OnComplete",
    "OnConnect",
    "OnDisconnect",
    "OnOperation",
    "Operation",
    "WebSocketConnectionError",
]
