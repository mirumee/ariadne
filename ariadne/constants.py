DATA_TYPE_JSON = "application/json"
DATA_TYPE_MULTIPART = "multipart/form-data"

CONTENT_TYPE_JSON = "application/json; charset=UTF-8"
CONTENT_TYPE_TEXT_HTML = "text/html; charset=UTF-8"
CONTENT_TYPE_TEXT_PLAIN = "text/plain; charset=UTF-8"

HTTP_STATUS_200_OK = "200 OK"
HTTP_STATUS_400_BAD_REQUEST = "400 Bad Request"
HTTP_STATUS_405_METHOD_NOT_ALLOWED = "405 Method Not Allowed"
