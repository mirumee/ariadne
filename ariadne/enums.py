import enum
from typing import (
    Any,
    Dict,
    Generator,
    List,
    Mapping,
    Optional,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
)
from functools import reduce, singledispatch
import operator

from graphql.type import GraphQLEnumType, GraphQLNamedType, GraphQLSchema
from graphql.language.ast import (
    EnumValueNode,
    InputValueDefinitionNode,
    ObjectValueNode,
)
from graphql.pyutils.undefined import Undefined
from graphql.type.definition import (
    GraphQLArgument,
    GraphQLField,
    GraphQLInputField,
    GraphQLInputObjectType,
    GraphQLInputType,
    GraphQLInterfaceType,
    GraphQLNonNull,
    GraphQLObjectType,
    GraphQLScalarType,
)

from .types import SchemaBindable


T = TypeVar("T")
ArgumentWithKeys = Tuple[str, str, GraphQLArgument, Optional[List["str"]]]
InputFieldWithKeys = Tuple[str, str, GraphQLInputField, Optional[List["str"]]]
GraphQLNamedInputType = Union[
    GraphQLScalarType, GraphQLEnumType, GraphQLInputObjectType
]


class EnumType(SchemaBindable):
    """Bindable mapping Python values to enumeration members in a GraphQL schema.

    # Example

    Given following GraphQL enum:

    ```graphql
    enum UserRole {
        MEMBER
        MODERATOR
        ADMIN
    }
    ```

    You can use `EnumType` to map it's members to Python `Enum`:

    ```python
    user_role_type = EnumType(
        "UserRole",
        {
            "MEMBER": 0,
            "MODERATOR": 1,
            "ADMIN": 2,
        }
    )
    ```

    `EnumType` also works with dictionaries:

    ```python
    user_role_type = EnumType(
        "UserRole",
        {
            "MEMBER": 0,
            "MODERATOR": 1,
            "ADMIN": 2,
        }
    )
    ```
    """

    def __init__(
        self,
        name: str,
        values: Union[Dict[str, Any], Type[enum.Enum], Type[enum.IntEnum]],
    ) -> None:
        """Initializes the `EnumType` with `name` and `values` mapping.

        # Required arguments

        `name`: a `str` with the name of GraphQL enum type in GraphQL schema to
        bind to.

        `values`: a `dict` or `enums.Enum` with values to use to represent GraphQL
        enum's in Python logic.
        """
        self.name = name
        self.values = cast(Dict[str, Any], getattr(values, "__members__", values))

    def bind_to_schema(self, schema: GraphQLSchema) -> None:
        """Binds this `EnumType` instance to the instance of GraphQL schema."""
        graphql_type = schema.type_map.get(self.name)
        self.validate_graphql_type(graphql_type)
        graphql_type = cast(GraphQLEnumType, graphql_type)

        for key, value in self.values.items():
            if key not in graphql_type.values:
                raise ValueError(
                    "Value %s is not defined on enum %s" % (key, self.name)
                )
            graphql_type.values[key].value = value

    def bind_to_default_values(self, schema: GraphQLSchema) -> None:
        """Populates default values of input fields and args in the GraphQL schema.

        This step is required because GraphQL query executor doesn't perform a
        lookup for default values defined in schema. Instead it simply pulls the
        value from fields and arguments `default_value` attribute, which is
        `None` by default.
        """
        for _, _, arg, key_list in find_enum_values_in_schema(schema):
            type_ = resolve_null_type(arg.type)
            type_ = cast(GraphQLNamedInputType, type_)

            if (
                key_list is None
                and arg.default_value in self.values
                and type_.name == self.name
            ):
                type_ = resolve_null_type(arg.type)
                arg.default_value = self.values[arg.default_value]

            elif key_list is not None:
                enum_value = get_value_from_mapping_value(arg.default_value, key_list)
                type_ = cast(GraphQLEnumType, track_type_for_nested(arg, key_list))

                if enum_value in self.values and type_.name == self.name:
                    set_leaf_value_in_mapping(
                        arg.default_value, key_list, self.values[enum_value]
                    )

    def validate_graphql_type(self, graphql_type: Optional[GraphQLNamedType]) -> None:
        """Validates that schema's GraphQL type associated with this `EnumType`
        is an `enum`."""
        if not graphql_type:
            raise ValueError("Enum %s is not defined in the schema" % self.name)
        if not isinstance(graphql_type, GraphQLEnumType):
            raise ValueError(
                "%s is defined in the schema, but it is instance of %s (expected %s)"
                % (self.name, type(graphql_type).__name__, GraphQLEnumType.__name__)
            )


def set_default_enum_values_on_schema(schema: GraphQLSchema):
    """Sets missing Python values for GraphQL enums in schema.

    Recursively scans GraphQL schema for enums and their values. If `value`
    attribute is empty, its populated with with a string of its GraphQL name.

    This string is then used to represent enum's value in Python instead of `None`.

    # Requires arguments

    `schema`: a GraphQL schema to set enums default values in.
    """
    for type_object in schema.type_map.values():
        if isinstance(type_object, GraphQLEnumType):
            set_default_enum_values(type_object)


def set_default_enum_values(graphql_type: GraphQLEnumType):
    for key in graphql_type.values:
        if graphql_type.values[key].value is None:
            graphql_type.values[key].value = key


def validate_schema_enum_values(schema: GraphQLSchema) -> None:
    """Raises `ValueError` if GraphQL schema has input fields or arguments with
    default values that are undefined enum values.

    # Example schema with invalid field argument

    This schema fails to validate because argument `role` on field `users`
    specifies `REVIEWER` as default value and `REVIEWER` is not a member of
    the `UserRole` enum:

    ```graphql
    type Query {
        users(role: UserRole = REVIEWER): [User!]!
    }

    enum UserRole {
        MEMBER
        MODERATOR
        ADMIN
    }

    type User {
        id: ID!
    }
    ```

    # Example schema with invalid input field

    This schema fails to validate because field `role` on input `UserFilters`
    specifies `REVIEWER` as default value and `REVIEWER` is not a member of
    the `UserRole` enum:

    ```graphql
    type Query {
        users(filter: UserFilters): [User!]!
    }

    input UserFilters {
        name: String
        role: UserRole = REVIEWER
    }

    enum UserRole {
        MEMBER
        MODERATOR
        ADMIN
    }

    type User {
        id: ID!
    }
    ```
    """

    for type_name, field_name, arg, _ in find_enum_values_in_schema(schema):
        if is_invalid_enum_value(arg):
            raise ValueError(
                f"Value for type: <{arg.type}> is invalid. "
                f"Check InputField/Arguments for <{field_name}> in <{type_name}> "
                "(Undefined enum value)."
            )


def is_invalid_enum_value(field: Union[GraphQLInputField, GraphQLArgument]) -> bool:
    if field.ast_node is None:
        return False
    return field.default_value is Undefined and field.ast_node.default_value is not None


def find_enum_values_in_schema(
    schema: GraphQLSchema,
) -> Generator[Union[ArgumentWithKeys, InputFieldWithKeys], None, None]:
    for name, type_ in schema.type_map.items():
        result = enum_values_in_types(type_, name)
        if result is not None:
            yield from result


@singledispatch
def enum_values_in_types(
    type_: GraphQLNamedType,  # pylint: disable=unused-argument
    name: str,  # pylint: disable=unused-argument
) -> Optional[Generator[Union[ArgumentWithKeys, InputFieldWithKeys], None, None]]:
    pass


@enum_values_in_types.register(GraphQLObjectType)
@enum_values_in_types.register(GraphQLInterfaceType)
def enum_values_in_object_type(
    type_: Union[GraphQLObjectType, GraphQLInterfaceType],
    field_name: str,
) -> Generator[ArgumentWithKeys, None, None]:
    for field in type_.fields.values():
        yield from enum_values_in_field_args(field_name, field)


@enum_values_in_types.register(GraphQLInputObjectType)
def enum_values_in_input_type(
    type_: GraphQLInputObjectType,
    field_name,
) -> Generator[InputFieldWithKeys, None, None]:
    yield from _get_field_with_keys(field_name, type_.fields.items())


def enum_values_in_field_args(
    field_name: str,
    field: GraphQLField,
) -> Generator[ArgumentWithKeys, None, None]:
    args = [
        (name, arg)
        for name, arg in field.args.items()
        if isinstance(
            arg.type, (GraphQLInputObjectType, GraphQLEnumType, GraphQLNonNull)
        )
    ]

    yield from _get_field_with_keys(field_name, args)


def _get_field_with_keys(field_name, fields):
    for input_name, field in fields:
        resolved_type = resolve_null_type(field.type)
        if isinstance(resolved_type, GraphQLEnumType):
            yield field_name, input_name, field, None

        if isinstance(resolved_type, GraphQLInputObjectType):
            if (
                field.ast_node is not None
                and field.ast_node.default_value is not None
                and isinstance(field.ast_node.default_value, ObjectValueNode)
            ):
                routes = get_enum_keys_from_ast(field.ast_node)
                for route in routes:
                    yield field_name, input_name, field, route


def get_enum_keys_from_ast(ast_node: InputValueDefinitionNode) -> List[List["str"]]:
    routes = []
    object_node = cast(ObjectValueNode, ast_node.default_value)
    nodes = [([field.name.value], field) for field in object_node.fields]

    while nodes:
        key_list, field = nodes.pop()
        if isinstance(field.value, EnumValueNode):
            routes.append(key_list)

        if isinstance(field.value, ObjectValueNode):
            for new_field in field.value.fields:
                new_route = key_list[:]
                new_route.append(new_field.name.value)
                nodes.append((new_route, new_field))

    return routes


def get_value_from_mapping_value(mapping: Mapping[T, Any], key_list: List[T]) -> Any:
    return reduce(operator.getitem, key_list, mapping)


def set_leaf_value_in_mapping(
    mapping: Mapping[T, Any], key_list: List[T], value: Any
) -> None:
    get_value_from_mapping_value(mapping, key_list[:-1])[key_list[-1]] = value


def track_type_for_nested(
    arg: Union[GraphQLArgument, GraphQLInputField], key_list: List[str]
) -> GraphQLInputType:
    type_ = resolve_null_type(arg.type)

    for elem in key_list:
        if isinstance(type_, GraphQLInputObjectType):
            type_ = type_.fields[elem].type
    return type_


def resolve_null_type(type_: GraphQLInputType) -> GraphQLInputType:
    return type_.of_type if isinstance(type_, GraphQLNonNull) else type_
