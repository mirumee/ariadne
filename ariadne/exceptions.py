from .constants import HTTP_STATUS_400_BAD_REQUEST


class HttpError(Exception):
    status = ""

    def __init__(self, message=None) -> None:
        super().__init__()
        self.message = message


class HttpBadRequestError(HttpError):
    status = HTTP_STATUS_400_BAD_REQUEST


class GraphQLFileSyntaxError(Exception):
    def __init__(self, schema_file, message) -> None:
        super().__init__()
        self.message = self.format_message(schema_file, message)

    def format_message(self, schema_file, message):
        return f"Could not load {schema_file}:\n{message}"

    def __str__(self):
        return self.message
