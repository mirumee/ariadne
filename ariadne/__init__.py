from .enums import (
    EnumType,
    set_default_enum_values_on_schema,
    validate_schema_enum_values,
)
from .executable_schema import make_executable_schema
from .extensions import ExtensionManager
from .file_uploads import combine_multipart_data, upload_scalar
from .format_error import (
    format_error,
    get_error_extension,
    get_formatted_error_context,
    get_formatted_error_traceback,
)
from .graphql import graphql, graphql_sync, subscribe
from .inputs import InputType
from .interfaces import InterfaceType, type_implements_interface
from .load_schema import load_schema_from_path
from .objects import MutationType, ObjectType, QueryType
from .resolvers import (
    FallbackResolversSetter,
    SnakeCaseFallbackResolversSetter,
    fallback_resolvers,
    is_default_resolver,
    resolve_to,
    snake_case_fallback_resolvers,
)
from .scalars import ScalarType
from .schema_names import SchemaNameConverter, convert_schema_names
from .schema_visitor import SchemaDirectiveVisitor
from .subscriptions import SubscriptionType
from .types import Extension, SchemaBindable
from .unions import UnionType
from .utils import (
    convert_camel_case_to_snake,
    convert_kwargs_to_snake_case,
    gql,
    unwrap_graphql_error,
)

__all__ = [
    "EnumType",
    "Extension",
    "ExtensionManager",
    "FallbackResolversSetter",
    "InputType",
    "InterfaceType",
    "MutationType",
    "ObjectType",
    "QueryType",
    "ScalarType",
    "SchemaBindable",
    "SchemaDirectiveVisitor",
    "SchemaNameConverter",
    "SnakeCaseFallbackResolversSetter",
    "SubscriptionType",
    "UnionType",
    "combine_multipart_data",
    "convert_camel_case_to_snake",
    "convert_kwargs_to_snake_case",
    "convert_schema_names",
    "fallback_resolvers",
    "format_error",
    "get_error_extension",
    "get_formatted_error_context",
    "get_formatted_error_traceback",
    "gql",
    "graphql",
    "graphql_sync",
    "is_default_resolver",
    "load_schema_from_path",
    "make_executable_schema",
    "resolve_to",
    "set_default_enum_values_on_schema",
    "snake_case_fallback_resolvers",
    "subscribe",
    "type_implements_interface",
    "unwrap_graphql_error",
    "upload_scalar",
    "validate_schema_enum_values",
]
