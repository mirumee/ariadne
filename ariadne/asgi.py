import asyncio
import json
from typing import (
    Any,
    AsyncGenerator,
    Callable,
    Dict,
    List,
    Optional,
    Union,
    Type,
    cast,
)

from graphql import GraphQLError, GraphQLSchema
from graphql.execution import MiddlewareManager
from starlette.datastructures import UploadFile
from starlette.requests import Request
from starlette.responses import HTMLResponse, JSONResponse, PlainTextResponse, Response
from starlette.types import Receive, Scope, Send
from starlette.websockets import WebSocket, WebSocketState, WebSocketDisconnect

from .constants import DATA_TYPE_JSON, DATA_TYPE_MULTIPART, PLAYGROUND_HTML
from .exceptions import HttpBadRequestError, HttpError
from .file_uploads import combine_multipart_data
from .format_error import format_error
from .graphql import graphql, subscribe
from .logger import log_error
from .types import ContextValue, ErrorFormatter, Extension, RootValue

GQL_CONNECTION_INIT = "connection_init"  # Client -> Server
GQL_CONNECTION_ACK = "connection_ack"  # Server -> Client
GQL_CONNECTION_ERROR = "connection_error"  # Server -> Client

# NOTE: The keep alive message type does not follow the standard due to connection optimizations
GQL_CONNECTION_KEEP_ALIVE = "ka"  # Server -> Client

GQL_CONNECTION_TERMINATE = "connection_terminate"  # Client -> Server
GQL_START = "start"  # Client -> Server
GQL_DATA = "data"  # Server -> Client
GQL_ERROR = "error"  # Server -> Client
GQL_COMPLETE = "complete"  # Server -> Client
GQL_STOP = "stop"  # Client -> Server

ExtensionList = Optional[List[Type[Extension]]]


class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        debug: bool = False,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        extensions: Union[Callable[[Any], ExtensionList], ExtensionList] = None,
        middleware: Optional[MiddlewareManager] = None,
        keepalive: float = None,
    ):
        self.context_value = context_value
        self.root_value = root_value
        self.debug = debug
        self.logger = logger
        self.error_formatter = error_formatter
        self.extensions = extensions
        self.middleware = middleware
        self.keepalive = keepalive
        self.schema = schema

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] == "http":
            await self.handle_http(scope=scope, receive=receive, send=send)
        elif scope["type"] == "websocket":
            await self.handle_websocket(scope=scope, receive=receive, send=send)
        else:
            raise ValueError("Unknown scope type: %r" % (scope["type"],))

    async def get_context_for_request(self, request: Any) -> Any:
        if callable(self.context_value):
            return self.context_value(request)
        return self.context_value or {"request": request}

    async def get_extensions_for_request(self, request: Any) -> ExtensionList:
        if callable(self.extensions):
            return self.extensions(request)
        return self.extensions

    async def handle_http(self, scope: Scope, receive: Receive, send: Send):
        request = Request(scope=scope, receive=receive)
        if request.method == "GET":
            response = await self.render_playground(request)
        elif request.method == "POST":
            response = await self.graphql_http_server(request)
        else:
            response = Response(status_code=405)
        await response(scope, receive, send)

    async def handle_websocket(self, scope: Scope, receive: Receive, send: Send):
        websocket = WebSocket(scope=scope, receive=receive, send=send)
        await self.websocket_server(websocket)

    async def render_playground(  # pylint: disable=unused-argument
        self, request: Request
    ) -> Response:
        return HTMLResponse(PLAYGROUND_HTML)

    async def graphql_http_server(self, request: Request) -> Response:
        try:
            data = await self.extract_data_from_request(request)
        except HttpError as error:
            return PlainTextResponse(error.message or error.status, status_code=400)

        context_value = await self.get_context_for_request(request)
        extensions = await self.get_extensions_for_request(request)

        success, response = await graphql(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
            extensions=extensions,
            middleware=self.middleware,
        )
        status_code = 200 if success else 400
        return JSONResponse(response, status_code=status_code)

    async def extract_data_from_request(self, request: Request):
        content_type = request.headers.get("Content-Type", "")
        content_type = content_type.split(";")[0]

        if content_type == DATA_TYPE_JSON:
            return await self.extract_data_from_json_request(request)
        if content_type == DATA_TYPE_MULTIPART:
            return await self.extract_data_from_multipart_request(request)

        raise HttpBadRequestError(
            "Posted content must be of type {} or {}".format(
                DATA_TYPE_JSON, DATA_TYPE_MULTIPART
            )
        )

    async def extract_data_from_json_request(self, request: Request):
        try:
            return await request.json()
        except (TypeError, ValueError):
            raise HttpBadRequestError("Request body is not a valid JSON")

    async def extract_data_from_multipart_request(self, request: Request):
        try:
            request_body = await request.form()
        except ValueError:
            raise HttpBadRequestError("Request body is not a valid multipart/form-data")

        try:
            operations = json.loads(request_body.get("operations"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'operations' multipart field is not a valid JSON"
            )
        try:
            files_map = json.loads(request_body.get("map"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'map' multipart field is not a valid JSON"
            )

        request_files = {
            key: value
            for key, value in request_body.items()
            if isinstance(value, UploadFile)
        }

        return combine_multipart_data(operations, files_map, request_files)

    async def websocket_server(self, websocket: WebSocket) -> None:
        subscriptions: Dict[str, AsyncGenerator] = {}
        await websocket.accept("graphql-ws")
        try:
            while websocket.application_state != WebSocketState.DISCONNECTED:
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
        except WebSocketDisconnect:
            pass
        finally:
            for operation_id in subscriptions:
                await subscriptions[operation_id].aclose()

    async def handle_websocket_message(
        self,
        message: dict,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        operation_id = cast(str, message.get("id"))
        message_type = cast(str, message.get("type"))

        if message_type == GQL_CONNECTION_INIT:
            await websocket.send_json({"type": GQL_CONNECTION_ACK})
            asyncio.ensure_future(self.keep_websocket_alive(websocket))
        elif message_type == GQL_CONNECTION_TERMINATE:
            await websocket.close()
        elif message_type == GQL_START:
            await self.start_websocket_subscription(
                message.get("payload"), operation_id, websocket, subscriptions
            )
        elif message_type == GQL_STOP:
            if operation_id in subscriptions:
                await subscriptions[operation_id].aclose()
                del subscriptions[operation_id]

    async def keep_websocket_alive(self, websocket: WebSocket):
        if not self.keepalive:
            return
        while websocket.application_state != WebSocketState.DISCONNECTED:
            try:
                await websocket.send_json({"type": GQL_CONNECTION_KEEP_ALIVE})
            except WebSocketDisconnect:
                return
            await asyncio.sleep(self.keepalive)

    async def start_websocket_subscription(
        self,
        data: Any,
        operation_id: str,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        context_value = await self.get_context_for_request(websocket)
        success, results = await subscribe(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
        )
        if not success:
            results = cast(List[dict], results)
            await websocket.send_json(
                {"type": GQL_ERROR, "id": operation_id, "payload": results[0]}
            )
        else:
            results = cast(AsyncGenerator, results)
            subscriptions[operation_id] = results
            asyncio.ensure_future(
                self.observe_async_results(results, operation_id, websocket)
            )

    async def observe_async_results(
        self, results: AsyncGenerator, operation_id: str, websocket: WebSocket
    ) -> None:
        try:
            async for result in results:
                payload = {}
                if result.data:
                    payload["data"] = result.data
                if result.errors:
                    for error in result.errors:
                        log_error(error, self.logger)
                    payload["errors"] = [
                        self.error_formatter(error, self.debug)
                        for error in result.errors
                    ]
                await websocket.send_json(
                    {"type": GQL_DATA, "id": operation_id, "payload": payload}
                )
        except Exception as error:
            if not isinstance(error, GraphQLError):
                error = GraphQLError(str(error), original_error=error)
            log_error(error, self.logger)
            payload = {"errors": [self.error_formatter(error, self.debug)]}
            await websocket.send_json(
                {"type": GQL_DATA, "id": operation_id, "payload": payload}
            )

        if websocket.application_state != WebSocketState.DISCONNECTED:
            await websocket.send_json({"type": GQL_COMPLETE, "id": operation_id})
