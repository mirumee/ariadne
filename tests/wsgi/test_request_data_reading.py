import json
from io import BytesIO

from ariadne.constants import HTTP_STATUS_400_BAD_REQUEST
from ariadne.exceptions import HttpBadRequestError

from .factories import create_multipart_request


def test_attempt_parse_request_missing_content_type_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory()
    request.pop("CONTENT_TYPE")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_parse_non_json_request_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory(content_type="text/plain")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_get_content_length_from_missing_header_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory()
    request.pop("CONTENT_LENGTH")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_get_content_length_from_malformed_header_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory(content_length="malformed")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_get_request_body_from_missing_wsgi_input_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory()
    request.pop("wsgi.input")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_get_request_body_from_empty_wsgi_input_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory()
    request["wsgi.input"] = BytesIO(b"")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_parse_non_json_request_body_raises_bad_request_error(
    middleware,
    start_response,
    snapshot,
    graphql_query_request_factory,
    error_response_headers,
):
    request = graphql_query_request_factory(raw_data="not-json")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, error_response_headers
    )
    snapshot.assert_match(result)


def test_attempt_parse_json_scalar_request_raises_graphql_bad_request_error(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(raw_data=json.dumps("json string"))
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_attempt_parse_json_array_request_raises_graphql_bad_request_error(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(raw_data=json.dumps([1, 2, 3]))
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HttpBadRequestError.status, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_multipart_form_request_fails_if_operations_is_not_valid_json(
    middleware, snapshot, start_response, error_response_headers
):
    data = """
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="operations"

not a json
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="map"

{ "0": ["variables.file"] }
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="0"; filename="test.txt"
Content-Type: text/plain

test

--------------------------cec8e8123c05ba25--
    """.strip().replace(
        "\n", "\r\n"
    )

    request = create_multipart_request(data)
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, error_response_headers
    )
    snapshot.assert_match(result)


def test_multipart_form_request_fails_if_map_is_not_valid_json(
    middleware, snapshot, start_response, error_response_headers
):
    data = """
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="operations"

{ "query": "mutation ($file: Upload!) { upload(file: $file) }", "variables": { "file": null } }
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="map"

not a json
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="0"; filename="test.txt"
Content-Type: text/plain

test

--------------------------cec8e8123c05ba25--
    """.strip().replace(
        "\n", "\r\n"
    )

    request = create_multipart_request(data)
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, error_response_headers
    )
    snapshot.assert_match(result)
