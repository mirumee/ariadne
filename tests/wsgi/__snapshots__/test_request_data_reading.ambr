# serializer version: 1
# name: test_attempt_get_content_length_from_malformed_header_raises_bad_request_error
  list([
    b'Content length header is missing or incorrect',
  ])
# ---
# name: test_attempt_get_content_length_from_missing_header_raises_bad_request_error
  list([
    b'Content length header is missing or incorrect',
  ])
# ---
# name: test_attempt_get_request_body_from_empty_wsgi_input_raises_bad_request_error
  list([
    b'Request body cannot be empty',
  ])
# ---
# name: test_attempt_get_request_body_from_missing_wsgi_input_raises_bad_request_error
  list([
    b'Request body cannot be empty',
  ])
# ---
# name: test_attempt_parse_json_array_request_raises_graphql_bad_request_error
  dict({
    'errors': list([
      dict({
        'message': 'Operation data should be a JSON object',
      }),
    ]),
  })
# ---
# name: test_attempt_parse_json_scalar_request_raises_graphql_bad_request_error
  dict({
    'errors': list([
      dict({
        'message': 'Operation data should be a JSON object',
      }),
    ]),
  })
# ---
# name: test_attempt_parse_non_json_request_body_raises_bad_request_error
  list([
    b'Request body is not a valid JSON',
  ])
# ---
# name: test_attempt_parse_non_json_request_raises_bad_request_error
  list([
    b'Posted content must be of type application/json or multipart/form-data',
  ])
# ---
# name: test_attempt_parse_request_missing_content_type_raises_bad_request_error
  list([
    b'Posted content must be of type application/json or multipart/form-data',
  ])
# ---
# name: test_multipart_form_request_fails_if_map_is_not_valid_json
  list([
    b"Request 'map' multipart field is not a valid JSON",
  ])
# ---
# name: test_multipart_form_request_fails_if_operations_is_not_valid_json
  list([
    b"Request 'operations' multipart field is not a valid JSON",
  ])
# ---
